# 
# Exploratory Data Analysis



### Table of content

- [Data Description Report](#DDR)
- [Droping columns with 100% missing values](#MV)
- [Exploratory Data Analysis:](#EDA)
  - [#Number of Physician in the dataset](#Q1)
  - [#Number of beneficiary in the dataset](#Q2)
  - [Does Physicians serve different Hospitals?](#Q3)
  - [Does beneficiaries go to different hospitals?](#Q4)
  - [#Number of Patient & Physician vs Potential Fraud](#Q5)
  - [Beneficiaries who visit the hospital more often have worse health conditions than the others?](#Q6)


# 
###  Import necessary packages

In [1]:
import pandas as pd
import pandas_profiling
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
plt.rcParams['figure.figsize'] = (12,8)
import warnings
warnings.filterwarnings("ignore")

# 
### Read Dataset

In [2]:
Beneficiary = pd.read_csv('Data/Train_Beneficiary.csv')
Inpatient = pd.read_csv('Data/Train_Inpatient.csv')
Outpatient = pd.read_csv('Data/Train_Outpatient.csv')
Label = pd.read_csv('Data/Train_Label.csv')

In [3]:
Label.head()

Provider PotentialFraud
0  PRV51001             No
1  PRV51003            Yes
2  PRV51004             No
3  PRV51005            Yes
4  PRV51007             No

In [4]:
display(Beneficiary.head())
display(Inpatient.head())
display(Outpatient.head())
display(Label.head())

BeneID         DOB  DOD  Gender  Race RenalDiseaseIndicator  State  \
0  BENE11001  1943-01-01  NaN       1     1                     0     39   
1  BENE11002  1936-09-01  NaN       2     1                     0     39   
2  BENE11003  1936-08-01  NaN       1     1                     0     52   
3  BENE11004  1922-07-01  NaN       1     1                     0     39   
4  BENE11005  1935-09-01  NaN       1     1                     0     24   

   County  NoOfMonths_PartACov  NoOfMonths_PartBCov  ...  \
0     230                   12                   12  ...   
1     280                   12                   12  ...   
2     590                   12                   12  ...   
3     270                   12                   12  ...   
4     680                   12                   12  ...   

   ChronicCond_Depression  ChronicCond_Diabetes  ChronicCond_IschemicHeart  \
0                       1                     1                          1   
1                       2                     2                          2   
2                       2                     2                          1   
3                       2                     1                          1   
4                       2                     1                          2   

   ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis  \
0                         2                                1   
1                         2                                2   
2                         2                                2   
3                         1                                1   
4                         2                                2   

   ChronicCond_stroke  IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  \
0                   1                     36000                   3204   
1                   2                         0                      0   
2                   2                         0                      0   
3                   2                         0                      0   
4                   2                         0                      0   

   OPAnnualReimbursementAmt  OPAnnualDeductibleAmt  
0                        60                     70  
1                        30                     50  
2                        90                     40  
3                      1810                    760  
4                      1790                   1200  

[5 rows x 25 columns]

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE11001  CLM66048   2009-08-31  2009-09-02  PRV55907   
2  BENE11001  CLM68358   2009-09-17  2009-09-20  PRV56046   
3  BENE11011  CLM38412   2009-02-14  2009-02-22  PRV52405   
4  BENE11014  CLM63689   2009-08-13  2009-08-30  PRV56614   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                    5000          PHY318495          PHY318495   
2                    5000          PHY372395                NaN   
3                    5000          PHY369659          PHY392961   
4                   10000          PHY379376          PHY398258   

  OtherPhysician AdmissionDt  ... ClmDiagnosisCode_7  ClmDiagnosisCode_8  \
0            NaN  2009-04-12  ...               2724               19889   
1            NaN  2009-08-31  ...                NaN                 NaN   
2      PHY324689  2009-09-17  ...                NaN                 NaN   
3      PHY349768  2009-02-14  ...              25062               40390   
4            NaN  2009-08-13  ...               5119               29620   

  ClmDiagnosisCode_9 ClmDiagnosisCode_10 ClmProcedureCode_1  \
0               5849                 NaN                NaN   
1                NaN                 NaN             7092.0   
2                NaN                 NaN                NaN   
3               4019                 NaN              331.0   
4              20300                 NaN             3893.0   

  ClmProcedureCode_2 ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  ClmProcedureCode_6  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 30 columns]

BeneID    ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11002  CLM624349   2009-10-11  2009-10-11  PRV56011   
1  BENE11003  CLM189947   2009-02-12  2009-02-12  PRV57610   
2  BENE11003  CLM438021   2009-06-27  2009-06-27  PRV57595   
3  BENE11004  CLM121801   2009-01-06  2009-01-06  PRV56011   
4  BENE11004  CLM150998   2009-01-22  2009-01-22  PRV56011   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                      30          PHY326117                NaN   
1                      80          PHY362868                NaN   
2                      10          PHY328821                NaN   
3                      40          PHY334319                NaN   
4                     200          PHY403831                NaN   

  OtherPhysician ClmDiagnosisCode_1  ... ClmDiagnosisCode_9  \
0            NaN              78943  ...                NaN   
1            NaN               6115  ...                NaN   
2            NaN               2723  ...                NaN   
3            NaN              71988  ...                NaN   
4            NaN              82382  ...                NaN   

  ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
0                 NaN                NaN                NaN   
1                 NaN                NaN                NaN   
2                 NaN                NaN                NaN   
3                 NaN                NaN                NaN   
4                 NaN                NaN                NaN   

  ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  DeductibleAmtPaid  ClmAdmitDiagnosisCode  
0                 0                  56409  
1                 0                  79380  
2                 0                    NaN  
3                 0                    NaN  
4                 0                  71947  

[5 rows x 27 columns]

Provider PotentialFraud
0  PRV51001             No
1  PRV51003            Yes
2  PRV51004             No
3  PRV51005            Yes
4  PRV51007             No

# 
###  Merge the 4 DataSets

#### First Combine Inpatient and Outpatient Data by BeneID

In [5]:
Inpatient['Patient_type'] = 'inpatient'
Outpatient['Patient_type'] = 'outpatient'

In [6]:
Patient = pd.concat([Inpatient, Outpatient], axis=0)
Patient.head()

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE11001  CLM66048   2009-08-31  2009-09-02  PRV55907   
2  BENE11001  CLM68358   2009-09-17  2009-09-20  PRV56046   
3  BENE11011  CLM38412   2009-02-14  2009-02-22  PRV52405   
4  BENE11014  CLM63689   2009-08-13  2009-08-30  PRV56614   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                    5000          PHY318495          PHY318495   
2                    5000          PHY372395                NaN   
3                    5000          PHY369659          PHY392961   
4                   10000          PHY379376          PHY398258   

  OtherPhysician AdmissionDt  ... ClmDiagnosisCode_8  ClmDiagnosisCode_9  \
0            NaN  2009-04-12  ...              19889                5849   
1            NaN  2009-08-31  ...                NaN                 NaN   
2      PHY324689  2009-09-17  ...                NaN                 NaN   
3      PHY349768  2009-02-14  ...              40390                4019   
4            NaN  2009-08-13  ...              29620               20300   

  ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
0                 NaN                NaN                NaN   
1                 NaN             7092.0                NaN   
2                 NaN                NaN                NaN   
3                 NaN              331.0                NaN   
4                 NaN             3893.0                NaN   

  ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  Patient_type  
0    inpatient  
1    inpatient  
2    inpatient  
3    inpatient  
4    inpatient  

[5 rows x 31 columns]

#### Finnaly, Combine Patient Data with Beneficiary Data and Labels 

In [7]:
fraud = pd.merge(Patient, Beneficiary, on='BeneID', how='left')
fraud = pd.merge(fraud, Label, on='Provider', how='left')
fraud.head()

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE11001  CLM66048   2009-08-31  2009-09-02  PRV55907   
2  BENE11001  CLM68358   2009-09-17  2009-09-20  PRV56046   
3  BENE11011  CLM38412   2009-02-14  2009-02-22  PRV52405   
4  BENE11014  CLM63689   2009-08-13  2009-08-30  PRV56614   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                    5000          PHY318495          PHY318495   
2                    5000          PHY372395                NaN   
3                    5000          PHY369659          PHY392961   
4                   10000          PHY379376          PHY398258   

  OtherPhysician AdmissionDt  ... ChronicCond_Diabetes  \
0            NaN  2009-04-12  ...                    1   
1            NaN  2009-08-31  ...                    1   
2      PHY324689  2009-09-17  ...                    1   
3      PHY349768  2009-02-14  ...                    1   
4            NaN  2009-08-13  ...                    2   

   ChronicCond_IschemicHeart ChronicCond_Osteoporasis  \
0                          1                        2   
1                          1                        2   
2                          1                        2   
3                          2                        2   
4                          1                        2   

  ChronicCond_rheumatoidarthritis ChronicCond_stroke IPAnnualReimbursementAmt  \
0                               1                  1                    36000   
1                               1                  1                    36000   
2                               1                  1                    36000   
3                               1                  1                     5000   
4                               2                  2                    21260   

  IPAnnualDeductibleAmt OPAnnualReimbursementAmt OPAnnualDeductibleAmt  \
0                  3204                       60                    70   
1                  3204                       60                    70   
2                  3204                       60                    70   
3                  1068                      250                   320   
4                  2136                      120                   100   

  PotentialFraud  
0            Yes  
1             No  
2             No  
3             No  
4             No  

[5 rows x 56 columns]

# 
### Data Exploration

#### Data Description Report :<a class="anchor" id="DDR"></a>
***First let's have on the features that we have***

### From the report we can see that : 

- ``` OperatingPhysician```  has <span style="color: red;">443764 (79.5%) missing values</span>
- ``` OtherPhysician```  has <span style="color: orange;">358475 (64.2%) missing values </span>
- ``` AdmissionDt```  has <span style="color: red;">517737 (92.7%) missing values</span>
- ``` ClmAdmitDiagnosisCode```  has <span style="color: red;">412312 (73.9%) missing values</span>
- ``` DischargeDt```  has <span style="color: red;">517737 (92.7%) missing values </span>
- ``` DiagnosisGroupCode``` has <span style="color: red;">517737 (92.7%) missing values </span>
- ``` ClmDiagnosisCode_1``` has 10453 (1.9%) missing values
- ``` ClmDiagnosisCode_2``` has 195606 (35.0%) missing values 
- ``` ClmDiagnosisCode_3``` has 315156 (56.5%) missing values 
- ``` ClmDiagnosisCode_4``` has <span style="color: orange;">393675 (70.5%) missing values </span>
- ``` ClmDiagnosisCode_5``` has <span style="color: orange;">446287 (79.9%) missing values </span>
- ``` ClmDiagnosisCode_6``` has <span style="color: red;">473819 (84.9%) missing values</span> 
- ``` ClmDiagnosisCode_7``` has <span style="color: red;">492034 (88.1%) missing values </span>
- ``` ClmDiagnosisCode_8``` has <span style="color: red;">504767 (90.4%) missing values </span>
- ``` ClmDiagnosisCode_9``` has <span style="color: red;">516396 (92.5%) missing values </span>
- ``` ClmDiagnosisCode_10```  has <span style="color: red;">553201 (99.1%) missing values </span>
- ``` ClmProcedureCode_1``` has <span style="color: red;">534901 (95.8%) missing values </span>
- ``` ClmProcedureCode_2``` has <span style="color: red;">552721 (99.0%) missing values </span>
- ``` ClmProcedureCode_3``` has <span style="color: red;">557242 (99.8%) missing values </span>
- ``` ClmProcedureCode_4``` has <span style="color: red;">558093 (> 99.9%) missing values </span>
- ``` ClmProcedureCode_5``` has <span style="color: red;">558202 (> 99.9%) missing values </span>
- ``` ClmProcedureCode_6``` has <span style="color: red;">558211 (100.0%) missing values </span>
- ``` DOD```  has <span style="color: red;">554080 (99.3%) missing values </span>

**Probably we gonna remove a lot of those**<br>
**Also we have a lot of features with ``Zero``, where we need to investigate more.**

# 
### Droping columns with 100% missing values<a class="anchor" id="MV"></a>

In [8]:
fraud = fraud.drop(['DOD'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_6'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_5'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_4'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_3'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_2'],axis = 1)
fraud = fraud.drop(['ClmProcedureCode_1'],axis = 1)
fraud = fraud.drop(['ClmDiagnosisCode_10'],axis = 1)
fraud = fraud.drop(['ClmDiagnosisCode_9'],axis = 1)
fraud = fraud.drop(['ClmDiagnosisCode_8'],axis = 1)

# 
## Let's check how many Beneficiary we have in our dataset<a class="anchor" id="EDA"></a>
### Let's check how many Beneficiary we have in our dataset<a class="anchor" id="Q1"></a>

In [9]:
print(fraud[fraud.Patient_type=='inpatient'].BeneID.unique().shape[0])
print(fraud[fraud.Patient_type=='outpatient'].BeneID.unique().shape[0])

31289
133980


**We have 31 289 inpatients and 133 980 outpatients**

### Also, Let's check how many doctors we have<a class="anchor" id="Q2"></a>

We have 3 type of Physician in our dataset: 
- Attending Physician
- Operating Physician
- OtherPhysician

In [10]:
Att_Physician = fraud[['Provider', 'AttendingPhysician']].rename(columns = {'AttendingPhysician': 'Physician'})
Op_Physician = fraud[['Provider', 'OperatingPhysician']].rename(columns = {'OperatingPhysician': 'Physician'})
Ot_Physician = fraud[['Provider', 'OtherPhysician']].rename(columns = {'OtherPhysician': 'Physician'})

physician = pd.concat([Att_Physician, Op_Physician,Ot_Physician], axis = 0)

In [11]:
physician.head()

Provider  Physician
0  PRV55912  PHY390922
1  PRV55907  PHY318495
2  PRV56046  PHY372395
3  PRV52405  PHY369659
4  PRV56614  PHY379376

In [12]:
physician.Physician.nunique()

100737

**We have 100737 unique Physician in our dataset**

### Let's check if our doctors serve for different hospital<a class="anchor" id="Q3"></a>

In [13]:
diff = physician.groupby(['Physician'])['Provider'].nunique().reset_index(name = 'multiprovider')
#diff = physician.groupby(['Physician','Provider'])['Physician'].count().reset_index(name = 'multiprovider')\
#.groupby(['Physician'])['multiprovider'].count().reset_index()

In [14]:
diff.sort_values(ascending = False, by = 'multiprovider')

Physician  multiprovider
25499   PHY342165             10
44249   PHY365132              8
86936   PHY417254              8
17405   PHY332314              6
64027   PHY389249              6
...           ...            ...
34923   PHY353702              1
34922   PHY353701              1
34920   PHY353699              1
34919   PHY353698              1
100736  PHY434158              1

[100737 rows x 2 columns]

In [15]:
diff.groupby('multiprovider')['Physician'].count().apply(lambda x: x*100/diff['multiprovider'].sum()).reset_index(name='% number_physician_provider')

multiprovider  % number_physician_provider
0              1                    85.104126
1              2                     6.364609
2              3                     0.549667
3              4                     0.074082
4              5                     0.027438
5              6                     0.010060
6              8                     0.001829
7             10                     0.000915

Yes, we have some doctors that serves different Hospitals

# 
### Let's check if our beneficiary goes to different hospital or not, and if we can caracterize them<a class="anchor" id="Q4"></a>

**outcome : We can use this as a caracteristic for patient. We can generate a new feature providers/patient.**

In [16]:
patient = fraud.groupby(['BeneID','PotentialFraud'])['Provider'].nunique().reset_index(name = 'NumProvider')

In [17]:
provider = fraud.groupby(['BeneID'])['Provider'].nunique().reset_index(name = 'NumProvider')

In [18]:
patient.sort_values(ascending = False, by = 'NumProvider')

BeneID PotentialFraud  NumProvider
31072   BENE122118             No           15
31106   BENE122143             No           14
44782   BENE131419             No           14
117857   BENE40202             No           13
194565   BENE97705             No           13
...            ...            ...          ...
117274   BENE39748            Yes            1
46310   BENE132494            Yes            1
117272   BENE39747            Yes            1
46312   BENE132495            Yes            1
0       BENE100000            Yes            1

[197624 rows x 3 columns]

In [19]:
# Build cepp
import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data
app = JupyterDash(__name__)
app.layout = html.Div([
    html.H1("Number of Provider per Patient"),
    dcc.Graph(id='graph'),
    html.Label([
        "Input Variable",
        dcc.Dropdown(
            id='column-dropdown', clearable=False,
            value = patient.columns[0], options=[
                {'label': c, 'value': c}
                for c in patient.columns
            ])
    ]),
])
# Define callback to update graph
@app.callback(
    Output('graph', 'figure'),
    [Input("column-dropdown", "value")]
)
def update_figure(column):
    return px.histogram(patient, x='NumProvider', title="Histogram Plot")
# Run app and display result inline in the notebook
app.run_server(mode='inline')

In [20]:
patients = fraud.groupby(['Provider', 'BeneID'])['BeneID'].count().reset_index(name='NumPatients').\
groupby('Provider')['NumPatients'].count().reset_index()

provider_per_patient= pd.merge(patients, Label, on='Provider', how='left')

In [21]:
# Build cepp
import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data
app = JupyterDash(__name__)
app.layout = html.Div([
    html.H1("Number of Provider per Beneficiary"),
    dcc.Graph(id='graph'),
    html.Label([
        "Input Variable",
        dcc.Dropdown(
            id='column-dropdown', clearable=False,
            value = provider_per_patient.columns[0], options=[
                {'label': c, 'value': c}
                for c in provider_per_patient.columns
            ])
    ]),
])
# Define callback to update graph
@app.callback(
    Output('graph', 'figure'),
    [Input("column-dropdown", "value")]
)
def update_figure(column):
    return px.box(provider_per_patient, x='PotentialFraud',y='NumPatients',points = False,color="PotentialFraud",  title="Box Plot")
# Run app and display result inline in the notebook
app.run_server(mode='inline')

In [22]:
doctors= fraud.groupby(['Provider', 'AttendingPhysician'])['AttendingPhysician'].count().\
reset_index(name='NumOfDoctors').groupby('Provider')['NumOfDoctors'].count().reset_index()

patients= fraud.groupby(['Provider','BeneID'])['BeneID'].count().reset_index(name='NumOfPatients').\
groupby('Provider')['NumOfPatients'].count().reset_index()

In [23]:
fraudv2 = pd.merge(fraud, doctors, on='Provider')
fraudv2 = pd.merge(fraudv2, patients, on='Provider')

In [24]:
display(fraudv2.head())
display(doctors.head())
display(patient.head())

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE17521  CLM34721   2009-01-20  2009-02-01  PRV55912   
2  BENE21718  CLM72336   2009-10-17  2009-11-04  PRV55912   
3  BENE22934  CLM73394   2009-10-25  2009-10-29  PRV55912   
4  BENE24402  CLM32911   2009-01-08  2009-01-12  PRV55912   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                   19000          PHY349293          PHY370861   
2                   17000          PHY334706          PHY334706   
3                   13000          PHY390614          PHY323689   
4                    3000          PHY380413          PHY432598   

  OtherPhysician AdmissionDt  ... ChronicCond_Osteoporasis  \
0            NaN  2009-04-12  ...                        2   
1      PHY363291  2009-01-20  ...                        2   
2            NaN  2009-10-17  ...                        1   
3      PHY363291  2009-10-25  ...                        1   
4            NaN  2009-01-08  ...                        2   

   ChronicCond_rheumatoidarthritis ChronicCond_stroke  \
0                                1                  1   
1                                2                  2   
2                                2                  2   
3                                1                  2   
4                                2                  2   

  IPAnnualReimbursementAmt IPAnnualDeductibleAmt OPAnnualReimbursementAmt  \
0                    36000                  3204                       60   
1                    19000                  1068                      100   
2                    17000                  1068                     1050   
3                    27000                  2136                      450   
4                     3000                  1068                     8900   

  OPAnnualDeductibleAmt PotentialFraud NumOfDoctors NumOfPatients  
0                    70            Yes           68           100  
1                    20            Yes           68           100  
2                   540            Yes           68           100  
3                   160            Yes           68           100  
4                   740            Yes           68           100  

[5 rows x 48 columns]

Provider  NumOfDoctors
0  PRV51001            14
1  PRV51003            44
2  PRV51004            38
3  PRV51005             6
4  PRV51007            10

BeneID PotentialFraud  NumProvider
0  BENE100000            Yes            1
1  BENE100001             No            4
2  BENE100001            Yes            2
3  BENE100002             No            7
4  BENE100002            Yes            4

In [25]:
patient.NumProvider.value_counts()

1     107180
2      49655
3      22193
4      10013
5       4536
6       2234
7       1001
8        469
9        199
10        78
11        41
12        15
13         7
14         2
15         1
Name: NumProvider, dtype: int64

**Impressive, there is quite a few patients who consult more than one hospital**

### 
### Let's check if the Number of patients and doctors can help to detect Potential Fraud<a class="anchor" id="Q5"></a>

In [26]:
# Build App
import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data
app = JupyterDash(__name__)
app.layout = html.Div([
    html.H1("JupyterDash Demo"),
    dcc.Graph(id='graph'),
    html.Label([
        "Input Variable",
        dcc.Dropdown(
            id='column-dropdown', clearable=False,
            value = fraudv2.columns[0], options=[
                {'label': c, 'value': c}
                for c in ['PotentialFraud']
            ])
    ]),
])
# Define callback to update graph
@app.callback(
    Output('graph', 'figure'),
    [Input("column-dropdown", "value")]
)
def update_figure_3(column):
    return px.scatter(fraudv2, x="NumOfDoctors", y="NumOfPatients",color='PotentialFraud', render_mode="webgl", title="Histogram Plot")
# Run app and display result inline in the notebook
app.run_server(mode='inline')


# 
### Let's check if those who have more visits to the providers have worse health conditions than the others<a class="anchor" id="Q6"></a>

First let's define our chronique diseases

In [27]:
Inpatient = fraudv2[fraudv2['Patient_type']== 'inpatient']
outpatient = fraudv2[fraudv2['Patient_type']== 'outpatient']

In [28]:
Inpatient['chronic_diseases_count'] = Inpatient.filter(like = 'ChronicCond').sum(axis = 1)
outpatient['chronic_diseases_count'] = outpatient.filter(like = 'ChronicCond').sum(axis = 0)

In [29]:
nb_visit = Inpatient.groupby('BeneID')['BeneID'].agg(['count'])
Inpatient_visit_Count = Inpatient.merge(nb_visit, how = 'inner', on = 'BeneID')
Inpatient_visit_Count = Inpatient_visit_Count.rename(columns = {'count': 'nb_visit'})

In [30]:
T = Inpatient_visit_Count.groupby('chronic_diseases_count')['nb_visit'].agg(['sum'])

In [31]:
import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data
app = JupyterDash(__name__)
app.layout = html.Div([
    html.H1("JupyterDash Demo"),
    dcc.Graph(id='graph'),
    html.Label([
        "Input Variable",
        dcc.Dropdown(
            id='column-dropdown', clearable=False,
            value = T.columns[0], options=[
                {'label': c, 'value': c}
                for c in T.columns
            ])
    ]),
])
# Define callback to update graph
@app.callback(
    Output('graph', 'figure'),
    [Input("column-dropdown", "value")]
)
def update_figure(column):
    return px.bar(T, x='sum', title="Histogram Plot")
# Run app and display result inline in the notebook
app.run_server(mode='inline')